contents/testing/writing-testable-code.md

```diff
@@ -0,0 +1,110 @@
+
+## Don't do actual work in constructors
+
+Because you want to keep yourself away from trouble? Just help yourself and make the constructors do no work.
+
+> A class under tests can have tens of tests. Each test instantiates a slightly different object graph and than applies some stimulus and asserts a response.
+...
+> Any work you do in a constructor needs to successfully navigate through in every test (not just the direct test, but also any related test which tries to instantiate your class indirectly as part of some larger object graph).
+
+In short, the constructor's job is to assign the dependencies to fields. And that's all.
+
+## Avoid Global States
+
+Global states may be tolerable at runtime if you only have one instance of your application.
+
+However, each test is a small instantiation of your application. But the global states persists from one test to another (e.g. a global configuration file).
+
+When tests fail, there is possibility it cannot be reproduced in isolation due to the global states. In other words, order of tests matter, which is bad.
+
+## Avoid Singletons
+
+Because singletons are global states in sheep's clothing! It's easy to understand because Singletons enforce their singletoness through private constructor and a global instance variable.
+
+## Avoid Static Methods
+
+Because if you build your application only with static methods, even if you are using Java and Class, it is not an OOP application but a procedural application.
+
+And there is *no way* to unit test procedure application. Why?
+
+> The key to testing is the presence of seams (places where you can divert the normal execution flow). Seams are essentially polymorphism. Seams are needed so that you can isolate the unit of test.
+
+However,
+
+> A leaf method such as Math.abs() is not a problem since the execution call graph ends there. But if you pick a method in a core of your application logic than everything behind the method becomes hard to test, since there is no way to insert test doubles (and there are no seams). Additionally it is really easy for a leaf method to stop being a leaf and than a method which was OK as static no longer is.
+
+## Favor composition over inheritance
+
+Because
+
+> At run-time you can not chose a different inheritance, but you can chose a different composition. This is important for tests as we want to test thing in isolation.
+
+Wrongly used inheritance hurts clutter the focus of test because you need to mock out the your parent classes' irrelevant dependencies, too.
+
+> Inheriting from AuthenticatedServlet will make your sub-class very hard to test since every test will have to mock out the authentication. And you have to do so to successfully navigate the super class. But what if AuthenticatedServlet inherits from DbTransactionServlet? (that gets so much harder)
+
+## Favor polymorphism over conditionals
```

Comment:
Usually polymorphism has inheritance. Since your previous point was "Favor composition over inheritance", does this mean we should "Favor composition over conditionals"? Maybe you could be abit more specific for these 2 points